go/toolchain/toolchains.bzl

```diff
@@ -0,0 +1,195 @@
+  
+  # The full set of allowed os and arch combinations for the go toolchain
+  # This is the set of targets allowed, of which the set of hosts is a strict subset
+  android_arm = struct(os=os_android, arch=arch_arm)
+  darwin_386 = struct(os=os_darwin, arch=arch_386)
+  darwin_amd64 = struct(os=os_darwin, arch=arch_amd64)
+  darwin_arm = struct(os=os_darwin, arch=arch_arm)
+  darwin_arm64 = struct(os=os_darwin, arch=arch_arm64)
+  dragonfly_amd64 = struct(os=os_dragonfly, arch=arch_amd64)
+  freebsd_386 = struct(os=os_freebsd, arch=arch_386)
+  freebsd_amd64 = struct(os=os_freebsd, arch=arch_amd64)
+  freebsd_arm = struct(os=os_freebsd, arch=arch_arm)
+  linux_386 = struct(os=os_linux, arch=arch_386)
+  linux_amd64 = struct(os=os_linux, arch=arch_amd64)
+  linux_arm = struct(os=os_linux, arch=arch_arm)
+  linux_arm64 = struct(os=os_linux, arch=arch_arm64)
+  linux_ppc64 = struct(os=os_linux, arch=arch_ppc64)
+  linux_ppc64le = struct(os=os_linux, arch=arch_ppc64le)
+  linux_mips = struct(os=os_linux, arch=arch_mips)
+  linux_mipsle = struct(os=os_linux, arch=arch_mipsle)
+  linux_mips64 = struct(os=os_linux, arch=arch_mips64)
+  linux_mips64le = struct(os=os_linux, arch=arch_mips64le)
+  netbsd_386 = struct(os=os_netbsd, arch=arch_386)
+  netbsd_amd64 = struct(os=os_netbsd, arch=arch_amd64)
+  netbsd_arm = struct(os=os_netbsd, arch=arch_arm)
+  openbsd_386 = struct(os=os_openbsd, arch=arch_386)
+  openbsd_amd64 = struct(os=os_openbsd, arch=arch_amd64)
+  openbsd_arm = struct(os=os_openbsd, arch=arch_arm)
+  plan9_386 = struct(os=os_plan9, arch=arch_386)
+  plan9_amd64 = struct(os=os_plan9, arch=arch_amd64)
+  solaris_amd64 = struct(os=os_solaris, arch=arch_amd64)
+  windows_386 = struct(os=os_windows, arch=arch_386)
+  windows_amd64 = struct(os=os_windows, arch=arch_amd64)
+  
+  # The set of acceptable hosts for each of the go versions, this is essentially the
+  # set of sdk's we know how to fetch
+  versions = [
+      struct(
+          semver = [1,8,3],
+          hosts = [darwin_amd64, linux_386, linux_amd64, windows_386, windows_amd64, freebsd_386, freebsd_amd64]
+      ),
+      struct(
+          semver = [1,8,2],
+          hosts = [darwin_amd64, linux_amd64],
+      ),
+      struct(
+          semver = [1,8,1],
+          hosts = [darwin_amd64, linux_amd64],
+      ),
+      struct(
+          semver = [1,8,0],
+          hosts = [darwin_amd64, linux_amd64],
+      ),
+      struct(
+          semver = [1,7,6],
+          hosts = [darwin_amd64, linux_386, linux_amd64, windows_386, windows_amd64, freebsd_386, freebsd_amd64]
+      ),
+      struct(
+          semver = [1,7,5],
+          hosts = [darwin_amd64, linux_amd64],
+      ),
+  ]
+  
+  # The set of allowed cross compilations
+  cross_targets = {
+      linux_amd64: [windows_amd64],
+      darwin_amd64: [linux_amd64],
+  }
+
+  # Use all the above information to generate all the possible toolchain's we might support
+  toolchains = []
```

Comment:
This is both really complicated and pretty cool, it will help me solidify my thinking around how this can be used. Thanks!